xen/arm: smmuv3: Add cache maintenance for non-coherent SMMU queues

firscity opened the pull request:
According to the Arm SMMUv3 spec (ARM IHI 0070), a system may have SMMU(s) that is/are non-coherent to the PE (processing element). In such cases, memory accesses from the PE should be either non-cached or be augmented with manual cache maintenance. SMMU cache coherency is reported by bit 4 (COHACC) of the SMMU_IDR0 register and is already present in the Xen driver. However, the current implementation is not aware of cache maintenance for memory that is shared between the PE and non-coherent SMMUs.

During Xen bring-up on a system with non-coherent SMMUs, the driver did not work properly - the SMMU was not functional and halted initialization at the very beginning due to a timeout while waiting for CMD_SYNC completion.

To properly handle such scenarios, add the non_coherent flag to the arm_smmu_queue struct. It is initialized using features reported by the SMMU HW and will be used for triggering cache clean/invalidate operations. This flag is not queue-specific (it is applicable to the whole SMMU), but adding it to arm_smmu_queue allows us to not change function signatures and simplify the patch (smmu->features, which contains the required flag, are not available in code parts that require cache maintenance).

firscity committed:
xen/arm: smmuv3: Add cache maintenance for non-coherent SMMU queues

According to the Arm SMMUv3 spec (ARM IHI 0070), a system may have
SMMU(s) that is/are non-coherent to the PE (processing element). In such
cases, memory accesses from the PE should be either non-cached or be
augmented with manual cache maintenance. SMMU cache coherency is reported
by bit 4 (COHACC) of the SMMU_IDR0 register and is already present in the
Xen driver. However, the current implementation is not aware of cache
maintenance for memory that is shared between the PE and non-coherent
SMMUs. It contains dmam_alloc_coherent() function, that is added during
Linux driver porting. But it is actually a wrapper for _xzalloc(), that
returns normal writeback memory (which is OK for coherent SMMUs).

During Xen bring-up on a system with non-coherent SMMUs, the driver did
not work properly - the SMMU was not functional and halted initialization
at the very beginning due to a timeout while waiting for CMD_SYNC
completion:

  (XEN) SMMUv3: /soc/iommu@fa000000: CMD_SYNC timeout
  (XEN) SMMUv3: /soc/iommu@fa000000: CMD_SYNC timeout

To properly handle such scenarios, add the non_coherent flag to the
arm_smmu_queue struct. It is initialized using features reported by the
SMMU HW and will be used for triggering cache clean/invalidate operations.
This flag is not queue-specific (it is applicable to the whole SMMU), but
adding it to arm_smmu_queue allows us to not change function signatures
and simplify the patch (smmu->features, which contains the required flag,
are not available in code parts that require cache maintenance).

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity commented:
Please take a look
@lorc @otyshchenko1 @LKomaryanskiy

LKomaryanskiy commented on xen/drivers/passthrough/arm/smmu-v3.c:
```diff
@@ -341,10 +341,14 @@ static void queue_write(__le64 *dst, u64 *src, size_t n_dwords)
-	queue_write(Q_ENT(q, q->llq.prod), ent, q->ent_dwords);
+	queue_write(q_addr, ent, q->ent_dwords);
+	if (q->non_coherent)
+		clean_dcache_va_range(q_addr, q->ent_dwords * sizeof(__le64));
```
Here should be `sizeof(*q_addr)`

firscity replied:
Thank you, fixed.

LKomaryanskiy commented on xen/drivers/passthrough/arm/smmu-v3.c:
```diff
@@ -1587,6 +1603,8 @@ static int arm_smmu_init_one_queue(struct arm_smmu_device *smmu,
 
 	q->llq.prod = q->llq.cons = 0;
+
+	q->non_coherent = !(smmu->features & ARM_SMMU_FEAT_COHERENCY);
```
nit: maybe here should be a newline, before return?

firscity replied:
Previously there was not newline before the return, but I also thought that it is better to have it there. Added.

LKomaryanskiy reviewed (commented):
Maybe it is worth mentioning in the commit message 2 additional points:
1) Current implementation of SMMU uses `dmam_alloc_coherent`, but actually it is `_xzalloc`, so memory allocated for SMMU operations is not coherent. Thus, we need to explicitly maintain the cache.
2) On your choice to insert or not the error while SMMU initialization and while SMMU operations into the commit message:
> (XEN) SMMUv3: /soc/iommu@fa000000: CMD_SYNC timeout
> (XEN) SMMUv3: /soc/iommu@fa000000: CMD_SYNC timeout

Also, please check using the `queue_full` & `queue_empty`. Whether or not they require cache maintenance?

firscity reviewed (commented):
@LKomaryanskiy thank you very much for review!

> Maybe it is worth mentioning in the commit message 2 additional points:

Added both, thank you. Please re-check the commit message.

> Also, please check using the queue_full & queue_empty. Whether or not they require cache maintenance?

Yeah, thank you. These functions are using indexes, that are located in memory and are written to registers (cons/prod_reg) by separate functions. Thus, no cache maintenance needed there.

lorc reviewed (commented):
Couple of observations:

1.

> This flag is not queue-specific (it is applicable to the whole SMMU), but
adding it to arm_smmu_queue allows us to not change function signatures
and simplify the patch (smmu->features, which contains the required flag,
are not available in code parts that require cache maintenance).

This is weak excuse. There is 50% chance that you'll be asked to redesign the driver to store  (and use) this property in SMMU context.

2. Constant `if (q->non_coherent)` checks looks like eye-sore. It is better to introduce helper static function that will do cache invalidation/cleaning for you if needed.

3. It is better not to use negated boolean variables. In other words `q->coherent` will be better.

LKomaryanskiy commented on xen/drivers/passthrough/arm/smmu-v3.h:
```diff
@@ -522,6 +522,13 @@ struct arm_smmu_queue {
 	u32 __iomem			*prod_reg;
 	u32 __iomem			*cons_reg;
+
+	/*
```
nit: just to be aligned with mainline version, maybe it is worth using a shorter version of the comment:
`/* Is the memory access coherent? */`

Like here:
https://patchew.org/Xen/6f4552aab3748ea3ad96d45affb8ce9146b557a4.1756922110.git.dmytro._5Ffirsov@epam.com/

LKomaryanskiy reviewed (approved):
Sorry for the delay, but I think it is time to merge the patch and remove the patch file from our product.

With a small nit, that can be omitted: `Reviewed-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>`.
Also, I think it is worth adding tags:
```
Reviewed-by: Julien Grall <jgrall@amazon.com>
Tested-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Bertrand Marquis <bertrand.marquis@arm.com>
```

Thank you a lot for the patch!

firscity commented:
@LKomaryanskiy thank you for raising it, I changed comment in the same way as it done in mainline, added your RB, but I have some concerns regarding adding mainline tags here. Code is different in mainline and our Xen, thus I would like to proceed without them if you don't mind.

I believe it is ready to be merged.
